corehq/apps/case_importer/do_import.py

```diff
@@ -38,8 +38,31 @@
 
 
 def do_import(spreadsheet, config, domain, task=None, record_form_callback=None):
-    importer = _TimedAndThrottledImporter(domain, config, task, record_form_callback)
-    return importer.do_import(spreadsheet)
+    has_domain_column = 'domain' in [c.lower() for c in spreadsheet.get_header_columns()]
+    if has_domain_column:
+        mirror_domains = DomainPermissionsMirror.mirror_domains(domain)
+        sub_domains = set()
+        results = _ImportResults()
+        for row_num, row in enumerate(spreadsheet.iter_row_dicts(), start=1):
+            if row_num == 1:
+                continue
+            sheet_domain = row.get('domain')
+            if sheet_domain != domain and sheet_domain not in mirror_domains:
+                err = exceptions.CaseRowError(column_name='domain')
+                err.title = _('Invalid domain')
+                err.message = _('Following rows contain invalid value for domain column.')
+                results.add_error(row_num, err)
+            else:
+                sub_domains.add(sheet_domain)
+        for sub_domain in sub_domains:
+            importer = _TimedAndThrottledImporter(sub_domain, config, task, record_form_callback)
+            importer.results = results
```

Comment:
I'd pass this in the constructor and update `_Importer.__init__` to only create a new `_ImportResults` if one isn't given, rather than always creating a new `_ImportResults` and then sometimes overwriting it.